google/cloud/bigtable/app_profile_config.h

```diff
@@ -56,7 +56,7 @@ class AppProfileConfig {
   }
 
   // NOLINT: accessors can (and should) be snake_case.
-  google::bigtable::admin::v2::CreateAppProfileRequest as_proto_move() {
+  google::bigtable::admin::v2::CreateAppProfileRequest as_proto_move() && {
```

Comment:
I think we want these to return `CreateAppProfileRequest&&` types. That will be more efficient. If the function is declared to return a value, such as `T`, then it will construct a new T to return. Sure, in this case it will move-construct that T, but a new T will be constructed, then returned to the caller. That returned object itself will be an rvalue, so the receiving side will then move-construct its own T, but that's two move constructions, whereas returning `T&&` would allow for a single move construction.

We can see how clang handles this looking at the return statements on lines 23 and 34 at https://godbolt.org/z/CVVn15

Also, returning T&& is similar to what std::optional does (https://en.cppreference.com/w/cpp/utility/optional/value)

... which makes me think, do we want to name this `as_proto()` and simply overload based on the rvalue-ness and const-ness of *this? That would also follow what std::optional::value does.

Thoughts?

Reply:
> I think we want these to return CreateAppProfileRequest&& types.

Done. Except in one place: `TableConfig::as_proto()` really does build the proto by moving its member variables into a new object and returning it.

> Also, returning T&& is similar to what std::optional does (https://en.cppreference.com/w/cpp/utility/optional/value)

Compelling.

> ... which makes me think, do we want to name this as_proto() and simply overload based on the rvalue-ness and const-ness of *this? That would also follow what std::optional::value does.

> Thoughts?

Both you and @dopiera had the same thought.  It is done. Note that this requires changing the `as_proto() const` to `as_proto() const&` because if one overload is ref-qualified all other overloads have to be ref-qualified (13.1 (2.3) in the N4140 draft).